src/pyenphase/envoy.py

```diff
@@ -63,14 +66,39 @@ async def authenticate(
         if self._firmware.version >= AwesomeVersion("7.0.0"):
             # Envoy firmware using new token authentication
             _LOGGER.debug("Authenticating to Envoy using token authentication")
-            if (
+            if token is not None:
+                self.auth = EnvoyTokenAuth(token=token)
+            elif (
                 self.auth is None
                 and username is not None
                 and password is not None
                 and self._firmware.serial is not None
             ):
                 self.auth = EnvoyTokenAuth(username, password, self._firmware.serial)
-                await self.auth.setup()
+
+        if self.auth is not None:
+            await self.auth.setup(self._client, self._host)
+        else:
+            _LOGGER.error(
+                "You must include a token or username/password to authenticate to the Envoy."
+            )
+
+    async def request(self, endpoint: str) -> None:
+        """Make a request to the Envoy."""
+        if self.auth is None:
+            raise Exception(
```

Comment:
```suggestion
            raise RuntimeError(
```

Maybe `RuntimeError` ?

Reply:
Probably would be nicer to give it a named envoy exception to include the auth state is wrong.